src/main/java/edu/ie3/util/io/FileIOUtils.java

```diff
@@ -141,6 +157,495 @@ public static BufferedWriter getBufferedWriter(
+   * @throws FileException If the input file name is null or the input path is not of a valid file
+   */
+  private static File validateInputFileName(final Path fileName) throws FileException {
+    if (fileName == null) {
+      throw new FileException("Input file name is null.");
+    }
+
+    File inputFile = new File(fileName.toString());
+
+    if (!inputFile.isFile()) {
+      throw new FileException("Input path '" + fileName.toString() + "' is not of a valid file.");
+    }
+
+    return inputFile;
+  }
+
+  /**
+   * Checks, if the output file path actually ends with '.tar.gz' or '.gz'. If not, creates a
+   * filename from the directory or file name. Also, checks if the target file already exists. If it
+   * does not exist, finally creates the file.
+   *
+   * @param dirOrFileName path of the directory or file that should be compressed (full path
+   *     required in case of a directory / full path + filename + file extension required in case of
+   *     a file!)
+   * @param outputFileName the optional output filename (full path + filename + file extension
+   *     required!) if different from the provided filename
+   * @param isDir set to true if trying to compress a directory. Set to false if compressing a
+   *     single file. if set to true, will archive the files before compression
+   * @return Output file
+   * @throws FileException If the archive already exists or unable to create the output file
+   */
+  private static File validateOutputFileName(
+      final String dirOrFileName, final String outputFileName, boolean isDir) throws FileException {
+    String finalOutputFileName = null;
+
+    if (isDir) {
+      finalOutputFileName = outputFileName.endsWith(TARGZ) ? outputFileName : dirOrFileName + TARGZ;
+    } else {
+      finalOutputFileName = outputFileName.endsWith(GZ) ? outputFileName : dirOrFileName + GZ;
+    }
+
+    File outputFile = new File(finalOutputFileName);
+    if (outputFile.exists())
+      throw new FileException("The target '" + finalOutputFileName + "' already exists.");
+
+    try {
+      if (!outputFile.createNewFile())
+        throw new FileException("Cannot create file '" + finalOutputFileName + "'.");
+    } catch (IOException e) {
+      throw new FileException("Cannot create file '" + finalOutputFileName + "'.", e);
+    }
+
+    return outputFile;
+  }
+
+  /**
+   * Extracts the given archive to a sub-directory with the same name that the archive has beneath
+   * the target directory.
+   *
+   * @param archive Compressed tarball archive to extract
+   * @param target Path to the target folder
+   * @return Path to the actual folder, where the content is extracted to
+   * @throws FileException If the archive is not in a well shape, the target folder doesn't meet the
+   *     requirements or the archive tries to impose harm by exploiting zip slip vulnerability
+   */
+  public static Path extractDir(Path archive, Path target) throws FileException {
```

Comment:
For the compression of directories you wrapped the results into a `CompletableFuture`. Wouldn't this be a good option here as well?

Reply:
Okay, got it. I will change the function `extractDir(Path, Path)` accordingly.